### ISSUE TITLE
Sorry, don't know how to check guest version of Virtualbox Guest Additions on this platform. Stopping installation.

### ISSUE THREAD

**Issue by ggada** [dotless-de/vagrant-vbguest#291]
On Elementary OS Loki that's based on Ubuntu xenial (16.04) this is not able to detect the guest additions. Works OK on vanilla Ubuntu 16.04

```
% uname -r
4.15.14-041514-generic
% lsb_release -cs
loki
% vagrant vbguest --status
Sorry, don't know how to check guest version of Virtualbox Guest Additions on this platform. Stopping installation.
```

**Comment by totallyGreg** [comment]
Not able to detect solaris either.  
```
uname -a
SunOS ops-center-vm 5.11 11.3 i86pc i386 i86pc
vagrant@ops-center-vm:~$ /usr/bin/VBoxControl -V
5.2.0_RC1r118201
vagrant@ops-center-vm:~$ 
```

**Comment by danowar2k** [comment]
I've read up on the code somewhat and have experience with Solaris 11 and installing VBoxGuestAdditions manually for a box.

If you could provide me some pointers for Ruby syntax and your hooks and function logic (perhaps via IRC or something like that) I would try to provide a Installer class for Solaris 11.

**Comment by ggada** [comment]
Hey Danowar,

Did you figure out where the OS detection logic is located? The steps for elementary OS should be the same as Ubuntu 16.04. I can test and put in a pull request if it's a simple change.

**Comment by danowar2k** [comment]
Long time, would the pull request still be appreciated?

**Comment by fnordfish** [comment]
Sure. As long as it's kind of a universal thing/not depending on a special base box. A quick search over at "app.vagrantup.com" didn't yield anything that looks like very generic. Anyways, if you feel like it's going to be more complicated, please consider releasing it as an own gem (vagrant plugin).

**Comment by fnordfish** [comment]
I see. The thing is, that you are executing your code past vbguest. You didn't provide vbguest with anything.
While your code seems to be fairly specific to your setup, I take it, that it's working for you. In that case, you probably don't need vbguest at all.
If you'd like it to be a vbguest feature, you'd need to utilize vbguest as described in my comment here: https://github.com/dotless-de/vagrant-vbguest/issues/291#issuecomment-385916364

If you'd like to get some help implementing this, I'd suggest we take this conversation out of this thread (eg [gitter](https://gitter.im/dotless-de/vagrant-vbguest) or email).

**Comment by danowar2k** [comment]
What I meant was:
How does the vbguest plugin check for the Guest Additions version on a given VM? My script checks from the host using VBoxManage. I was wondering how the plugin does it. Does it enter the VM?
Just for understanding...

**Comment by fnordfish** [comment]
It does two things:

1. It asks Vagrant. Vagrant itself will call Virtualbox (IIRC, that's VBoxManage)
2. If vbguest finds a VBoxService inside your box, vbguest will ask this for it's version.

This is b/c VirtualBox is/was buggy and sometimes didn't update some internal metadata files on your host.

The key point here is, that vbguest offloads that kind of stuff to Vagrant so that we don't have to deal with it. BTW, same thing for detecting the main flavor of the guest os.

**Comment by fnordfish** [comment]
.... However, you'll keep seeing the "Sorry, don't know how to check guest version of Virtualbox Guest Additions on this platform" message b/c vbguest split up basically every work into `Installer` classes which are guest-os-specific. So even before vbguest tries to figure out the Virtualbox versions, it will try to find an Installer implementation for the guest system - and if it doesn't find one it'll just do nothing (besides printing that notice).

**Comment by danowar2k** [comment]
Hm. Wouldn't it be more logical to check for the Virtualbox/Guest Additions versions first if that is universally possible, and then check if there is an Installer for the OS?

Or alternatively to change the message to "Sorry, I wouldn't know how to install or update the Virtualbox Guest Additions on this platform"?

**Comment by fnordfish** [comment]
You are right, your suggested error message sound a lot better. Currently the error messages reflects in which actual call it got triggered (grep for `NoInstallerFoundError` in https://github.com/dotless-de/vagrant-vbguest/blob/master/lib/vagrant-vbguest/installer.rb). Maybe this granularity doesn't serve any purpose and things can be simplified.

**Comment by danowar2k** [comment]
By the way, if I do a pull request and include additional comments from when I try to understand Vagrant plugin structure and this plugin, is that ok?

**Comment by fnordfish** [comment]
I don’t mind comments. If it’s too excessive I might ask to strip it down, but don’t worry 😉

**Comment by fnordfish** [comment]
Released as 0.16.0

**Comment by danowar2k** [comment]
Still intending to work on this, still working on some other things that have to be finished first (Packer image for Solaris 11.4, Ansible playbooks for Redmine, Jenkins and other things...)
